### PR TITLE
[SPARK-50132][SQL][PYTHON] Add DataFrame API for Lateral Joins

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Adds DataFrame API for Lateral Joins.

#### Examples:

For the following DataFrames `customers` and `orders`:

```py
>>> customers.printSchema()
root
 |-- customer_id: long (nullable = true)
 |-- name: string (nullable = true)

>>> orders.printSchema()
root
 |-- order_id: long (nullable = true)
 |-- customer_id: long (nullable = true)
 |-- order_date: string (nullable = true)
 |-- items: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- product: string (nullable = true)
 |    |    |-- quantity: long (nullable = true)
```

##### Using TVF

```py
# select customer_id, name, order_id, order_date, product, quantity
# from customers join orders using (customer_id) join lateral (select col.* from explode(items))
# order by customer_id, order_id, product
customers.join(orders, "customer_id").lateralJoin(
    spark.tvf.explode(sf.col("items").outer()).select("col.*")
).select(
    "customer_id", "name", "order_id", "order_date", "product", "quantity"
).orderBy("customer_id", "order_id", "product").show()
```

##### Using Subquery

```py
# select c.customer_id, name, order_id, order_date
# from customers c left join lateral (
#     select * from orders o where o.customer_id = c.customer_id order by order_date desc limit 2
# )
# order by customer_id, order_id
customers.alias("c").lateralJoin(
    orders.alias("o")
    .where(sf.col("o.customer_id") == sf.col("c.customer_id").outer())
    .orderBy(sf.col("order_date").desc())
    .limit(2),
    how="left"
).select(
    "c.customer_id", "name", "order_id", "order_date"
).orderBy("customer_id", "order_id").show()
```

### Why are the changes needed?

Lateral Join APIs are missing in DataFrame API.

### Does this PR introduce _any_ user-facing change?

Yes, new DataFrame APIs for lateral join will be available.

### How was this patch tested?

Added the related tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -2629,6 +2629,103 @@ def join(
...
         ...
 
+    def lateralJoin(
+        self,
+        other: "DataFrame",
+        on: Optional[Column] = None,
+        how: Optional[str] = None,
+    ) -> "DataFrame":
+        """
+        Lateral joins with another :class:`DataFrame`, using the given join expression.
```
Comment (allisonwang-db) [inline python/pyspark/sql/dataframe.py:2639]:
```suggestion
        Lateral joins with another :class:`DataFrame`, using the given join expression.
        
        A lateral join (also known as a correlated join) is a type of join where each row from one DataFrame is used as input to a subquery or a derived table that computes a result specific to that row. The right side `DataFrame` can reference columns from the current row of the left side`DataFrame`, allowing for more complex and context-dependent results than a standard join.
```

**Review by allisonwang-db — APPROVED** [review]
Awesome!

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -383,6 +383,21 @@ class Dataset[T] private[sql] (
     }
   }
 
+  // TODO(SPARK-50134): Support Lateral Join API in Spark Connect
```
Comment (dongjoon-hyun) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:386]: According to the recent discussion on `Spark Connect`, is this considered a blocker issue for Apache Spark 4.0.0, @ueshin , @hvanhovell , and @HyukjinKwon ?

I believe this newly added API incompatibility (SPARK-50134) is a nice-to-have, and should be a blocker issue for Apache Spark 4.0.0.

Comment (HyukjinKwon) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:386]: IMO I think so ... but I think we should be fine in general ... as usually API are implemented in both Spark Classic and Connect at the same PR or at the similar timeline.

**Review by zhengruifeng — APPROVED** [review]

**Review by xinrong-meng — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
Could you re-trigger the failed CIs?

**Comment by ueshin** [comment]
@dongjoon-hyun Sure, retriggered [here](https://github.com/ueshin/apache-spark/actions/runs/12150631150/job/33936515254). 
But my GHA can't pass "Run / Build modules: pyspark-mllib, pyspark-ml, pyspark-ml-connect" job for some reason. I may be missing config or something?
I don't think it's related to this PR anyway.

**Comment by dongjoon-hyun** [comment]
> I don't think it's related to this PR anyway.

It seems that the last test failed due to timeout. 

- https://github.com/ueshin/apache-spark/actions/runs/12170708809/job/33946289331
Screenshot 2024-12-05 at 07 52 04

@ueshin . Please add more empty commits to see if that's CI issue or not.

**Comment by HyukjinKwon** [comment]
Ah, this specific test fails in forked repositories for some reasons.

**Comment by HyukjinKwon** [comment]
Let's merge, and see how it goes. I will monitor the build

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**